packages/dtslint/src/rules/npm-naming.ts

```diff
@@ -0,0 +1,233 @@
+
+type NpmNamingOptions = NameOnlyNpmNamingOptions | CodeNpmNamingOptions;
+
+interface NameOnlyNpmNamingOptions {
+  mode: "name-only";
+  "single-line"?: boolean;
+}
+
+interface CodeNpmNamingOptions {
+  mode: "code";
+  errors: [NpmNamingErrorKing, boolean][];
+  "single-line"?: boolean;
+}
+
+type NpmNamingErrorKing =
+  | "NeedsExportEquals"
+  | "NoDefaultExport"
+  | "JsPropertyNotInDts"
+  | "DtsPropertyNotInJs"
+  | "JsSignatureNotInDts"
+  | "DtsSignatureNotInJs";
+
+const messageMultiLine = `{{ explanation }}
+If you won't fix this error now or you think this error is wrong,
+you can disable this check by adding the following options to your project's .eslintrc.json file under "rules":
+
+    "npm-naming": {{ disable }}
+`;
+
+const rule = createRule<[NpmNamingOptions?], "multiLine" | "singleLine">({
+  name: "npm-naming",
+  defaultOptions: [],
+  meta: {
+    type: "problem",
+    docs: {
+      description: "Ensure that package name and DefinitelyTyped header match npm package info.",
+      recommended: "error",
+    },
+    messages: {
+      multiLine: messageMultiLine,
+      singleLine: toSingleLine(messageMultiLine),
+    },
+    schema: [
+      {
+        oneOf: [
+          {
+            type: "object",
+            properties: {
+              mode: {
+                type: "string",
+                enum: [Mode.NameOnly],
+              },
+              "single-line": {
+                description: "Whether to print error messages in a single line. Used for testing.",
+                type: "boolean",
+              },
+            },
+            required: ["mode"],
+          },
+          {
+            type: "object",
+            properties: {
+              mode: {
+                type: "string",
+                enum: [Mode.Code],
+              },
+              errors: {
+                type: "array",
+                items: {
+                  type: "array",
+                  items: [
+                    {
+                      description: "Name of the check.",
+                      type: "string",
+                      enum: [ErrorKind.NeedsExportEquals, ErrorKind.NoDefaultExport] as ExportErrorKind[],
+                    },
+                    {
+                      description: "Whether the check is enabled or disabled.",
+                      type: "boolean",
+                    },
+                  ],
+                  minItems: 2,
+                  maxItems: 2,
+                },
+              },
+              "single-line": {
+                description: "Whether to print error messages in a single line. Used for testing.",
+                type: "boolean",
+              },
+            },
+            required: ["mode"],
+          },
+        ],
+      },
+    ],
+  },
+  // TODO: This code is a modified version of the old TSLint rule,
+  // and still primarily uses TypeScript nodes. It would be better
+  // to switch it to using TSESTree nodes like other ESLint rules.
```

Comment:
> What are the drawbacks of using the entire TS AST? Is it slower? Harder to maintain?

There actually aren't that many drawbacks.
* It's not slower; `@typescript-eslint/parser` always creates both the ES tree and TS tree anyway. 
* It might be harder to maintain: ESLint rule authors are typically used to working on the ES tree